tap_s3_csv/conversion.py

```diff
@@ -1,98 +1,86 @@
+
 LOGGER = singer.get_logger()
 
-def infer(datum):
+
+def generate_schema(samples: List[Dict], table_spec: Dict) -> Dict:
     """
-    Returns the inferred data type
+    Guess columns types from the given samples and build json schema
+    :param samples: List of dictionaries containing samples data from csv file(s)
+    :param table_spec: table/stream specs given in the tap definition
+    :return: dictionary where the keys are the headers and values are the guessed types - compatible with json schema
     """
-    if datum is None or datum == '':
-        return None
+    counts = {}
 
-    try:
-        int(datum)
-        return 'integer'
-    except (ValueError, TypeError):
-        pass
+    table_set = CSVTableSet(_csv2bytesio(samples))
 
-    try:
-        #numbers are NOT floats, they are DECIMALS
-        float(datum)
-        return 'number'
-    except (ValueError, TypeError):
-        pass
+    row_set = table_set.tables[0]
 
-    return 'string'
+    offset, headers = headers_guess(row_set.sample)
+    row_set.register_processor(headers_processor(headers))
+    row_set.register_processor(offset_processor(offset + 1))
 
+    types = type_guess(row_set.sample, strict=True)
 
-def count_sample(sample, counts, table_spec):
-    for key, value in sample.items():
-        if key not in counts:
-            counts[key] = {}
+    for header, header_type in zip(headers, types):
 
-        date_overrides = table_spec.get('date_overrides', [])
-        if key in date_overrides:
-            datatype = "date-time"
-        else:
-            datatype = infer(value)
+        date_overrides = set(table_spec.get('date_overrides', []))
 
-        if datatype is not None:
-            counts[key][datatype] = counts[key].get(datatype, 0) + 1
+        if header in date_overrides:
+            counts[header] = {
+                'anyOf': [
+                    {'type': ['null', 'string'], 'format': 'date-time'},
+                    {'type': ['null', 'string']}
+                ]
+            }
+        else:
+            if isinstance(header_type, (DateType, DateUtilType)):
+                counts[header] = {
+                    'anyOf': [
+                        {'type': ['null', 'string'], 'format': 'date'},
+                        {'type': ['null', 'string']}
+                    ]
+                }
+            else:
+                counts[header] = {
+                    'type': ['null', 'string']
+                }
+
+                if isinstance(header_type, IntegerType):
+                    counts[header]['type'].append('integer')
+                elif isinstance(header_type, DecimalType):
+                    counts[header]['type'].append('number')
+                elif isinstance(header_type, BoolType):
+                    counts[header]['type'].append('boolean')
 
     return counts
 
 
-def pick_datatype(counts):
+def _csv2bytesio(data: List[Dict])-> io.BytesIO:
```

Comment:
I think it's always a good practice to use context managers whenever possible. io.StringIO() in this case. Here is some link: https://stackoverflow.com/questions/9718950/do-i-have-to-do-stringio-close

Reply:
Good catch! will adjust it.